isolate repo fork command and tweak usage

vilmibm opened the pull request:
Depends on #1421 
Part of #1405 
Part of #1082 

This PR is another isolation refactor, this time for repo fork.

However, I got fed up with the `--remote="true|false|prompt"` style of flags and took this opportunity to switch to bool flags. I didn't do this initially because of either a perceived or since-patched shortcoming of cobra's bool flags where it was impossible to tell the difference between a "false" flag and a flag that simply wasn't passed.

The interactions goes like this, now, assuming the command is attached to a TTY:

```
$ gh repo fork foo/bar --clone       # just clone, no prompt
$ gh repo fork foo/bar --clone=false # do not clone, no prompt
$ gh repo fork foo/bar                # prompt about cloning
```

vilmibm committed:
isolate repo fork command and tweak usage

This commit is another isolation refactor, this time for repo fork.
However, I got fed up with the --remote="true|false|prompt" style of
flags and took this opportunity to switch to a set of bool flags:

--remote and --clone
--no-remote and --no-clone

the string args were really non standard and confusing; with only two
bools it was impossible to tell when to prompt.

vilmibm commented:
pinging @ampinsk since I changed how the flags for this command work

mislav commented on pkg/cmd/repo/fork/fork.go:
```diff
@@ -0,0 +1,292 @@
+				opts.PromptRemote = true
+			}
+
+			if cmd.Flags().Changed("no-clone") {
```
Instead of `Changed`, could we look at the actual value of the `--no-clone` flag? Just in case someone tried to do `--no-clone=false`, which is strange but perfectly valid usage

vilmibm replied:
so...

I think I have been confused. I honestly didn't realize that cobra bool flags could be passed like `--clone=false` and I think instead of opening up the `-(no)-clone` can of worms I'd rather just do that for now.

The initial problem was that I thought cobra couldn't distinguish between:

- no flags specified (ie, prompt)
- desire is to set flag to false (which i thought could only be done by omitting the flag)

but since `--clone=false` works, I *can* identify the "desire for false" vs. the "no flags specified" case :facepalm: I'm going to retool for that.

vilmibm replied:
I switched to --clone/--remote as bools and things are fine :+1:

mislav commented on pkg/cmd/repo/fork/fork.go:
```diff
@@ -0,0 +1,292 @@
+	cmd.Flags().BoolVar(&opts.Clone, "clone", false, "Clone the fork")
+	cmd.Flags().BoolVar(&opts.Remote, "remote", false, "Add remote for fork")
+	cmd.Flags().Bool("no-clone", false, "Do not clone the fork")
+	cmd.Flags().Bool("no-remote", false, "Do not add remote for fork")
```
I wish that pflag had native support for `--no-something` flag format so these flags can be documented in relation to their originals rather than being listed as separate flags
```
--(no-)clone    Clone the fork
--(no-)remote   Add remote for fork
```
Maybe we can explore adding something like that to our docs as follow up? We might be able to pull it off specifically for our project without having to fork pflag

mislav commented on pkg/cmd/repo/fork/fork.go:
```diff
@@ -0,0 +1,292 @@
+	PromptRemote bool
+}
+
+// TODO decide if this should be injected via a factory.
+var Since = func(t time.Time) time.Duration {
+	return time.Since(t)
```
I don't think there is need to allow passing Since via factory just yet, as we can easily stub it via ForkOptions for tests where it matters.

However, when it comes to `ForkOptions.Since`, I wonder whether it would make sense to name it `ForkOptions.Now` with the value of `time.now()` at runtime and a fixed time in tests? We solve the same problem of freezing time, but in a more general way

vilmibm replied:
I specifically wanted to use the standard library's `Since` function and I would prefer to stub that out instead of stubbing `Now` and doing the math ourselves.

The field in `ForkOptions` is actually vestigial -- since we're just porting the existing test there is no split between testing argument processing and testing behavior.

For now I'm going to take out the `ForkOptions` field and when we eventually re-do the tests for fork we can revisit.

mislav reviewed (approved):
This is great 🚀

vilmibm committed:
just use simple bool flags

the original complexities here were because I thought there was a
shortcoming in cobra's bool support and I was wrong.

vilmibm committed:
remove vestigial struct field

vilmibm committed:
clarify usage
